In [1]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

import warnings
warnings.filterwarnings("ignore")

In [2]:
!mkdir -p ~/.postgresql && \
wget "https://storage.yandexcloud.net/cloud-certs/CA.pem" \
    --output-document ~/.postgresql/root.crt && \
chmod 0600 ~/.postgresql/root.crt

--2023-03-07 07:35:14--  https://storage.yandexcloud.net/cloud-certs/CA.pem
Resolving storage.yandexcloud.net (storage.yandexcloud.net)... 213.180.193.243, 2a02:6b8::1d9
Connecting to storage.yandexcloud.net (storage.yandexcloud.net)|213.180.193.243|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3579 (3.5K) [application/x-x509-ca-cert]
Saving to: ‘/root/.postgresql/root.crt’

/root/.postgresql/r 100%[===================>]   3.50K  --.-KB/s    in 0s      

2023-03-07 07:35:14 (569 MB/s) - ‘/root/.postgresql/root.crt’ saved [3579/3579]



In [3]:
!sudo apt update && sudo apt install --yes python3 python3-pip && \
!pip3 install psycopg2-binary

Get:1 https://cloud.r-project.org/bin/linux/ubuntu focal-cran40/ InRelease [3,622 B]
Hit:2 http://archive.ubuntu.com/ubuntu focal InRelease
Get:3 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu focal InRelease [18.1 kB]
Get:4 http://archive.ubuntu.com/ubuntu focal-updates InRelease [114 kB]
Get:5 http://security.ubuntu.com/ubuntu focal-security InRelease [114 kB]
Ign:6 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu2004/x86_64  InRelease
Hit:7 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64  InRelease
Hit:8 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu2004/x86_64  Release
Hit:9 http://ppa.launchpad.net/cran/libgit2/ubuntu focal InRelease
Get:10 http://archive.ubuntu.com/ubuntu focal-backports InRelease [108 kB]
Hit:11 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu focal InRelease
Hit:12 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu focal InRelease
Hit:13 http://ppa.launchpad.net/ubuntugis/p

In [13]:
import psycopg2

conn = psycopg2.connect("""
    host=rc1b-70q90cunwkzx49mj.mdb.yandexcloud.net
    port=6432
    sslmode=verify-full
    dbname=db_clients
    user=mlhs_student
    password=mlhs_student
    target_session_attrs=read-write
""")

In [14]:
import pandas as pd

query = "SELECT * from adult_income"
data = pd.read_sql_query(query, conn)

conn.close()

In [20]:
data.head(20)

age         workclass  fnlwgt     education  educational-num  \
0    25           Private  226802          11th                7   
1    38           Private   89814       HS-grad                9   
2    28         Local-gov  336951    Assoc-acdm               12   
3    44           Private  160323  Some-college               10   
4    18           Private  103497  Some-college               10   
5    34           Private  198693          10th                6   
6    29           Private  227026       HS-grad                9   
7    63  Self-emp-not-inc  104626   Prof-school               15   
8    24           Private  369667  Some-college               10   
9    55           Private  104996       7th-8th                4   
10   65           Private  184454       HS-grad                9   
11   36       Federal-gov  212465     Bachelors               13   
12   26           Private   82091       HS-grad                9   
13   58           Private  299831       HS-grad                9   
14   48           Private  279724       HS-grad                9   
15   43           Private  346189       Masters               14   
16   20         State-gov  444554  Some-college               10   
17   43           Private  128354       HS-grad                9   
18   37           Private   60548       HS-grad                9   
19   40           Private   85019     Doctorate               16   

        marital-status         occupation   relationship                race  \
0        Never-married  Machine-op-inspct      Own-child               Black   
1   Married-civ-spouse    Farming-fishing        Husband               White   
2   Married-civ-spouse    Protective-serv        Husband               White   
3   Married-civ-spouse  Machine-op-inspct        Husband               Black   
4        Never-married                  ?      Own-child               White   
5        Never-married      Other-service  Not-in-family               White   
6        Never-married                  ?      Unmarried               Black   
7   Married-civ-spouse     Prof-specialty        Husband               White   
8        Never-married      Other-service      Unmarried               White   
9   Married-civ-spouse       Craft-repair        Husband               White   
10  Married-civ-spouse  Machine-op-inspct        Husband               White   
11  Married-civ-spouse       Adm-clerical        Husband               White   
12       Never-married       Adm-clerical  Not-in-family               White   
13  Married-civ-spouse                  ?        Husband               White   
14  Married-civ-spouse  Machine-op-inspct        Husband               White   
15  Married-civ-spouse    Exec-managerial        Husband               White   
16       Never-married      Other-service      Own-child               White   
17  Married-civ-spouse       Adm-clerical           Wife               White   
18             Widowed  Machine-op-inspct      Unmarried               White   
19  Married-civ-spouse     Prof-specialty        Husband  Asian-Pac-Islander   

    gender  capital-gain  capital-loss  hours-per-week native-country income  
0     Male           0.0           0.0            40.0  United-States  <=50K  
1     Male           0.0           0.0            50.0  United-States  <=50K  
2     Male           0.0           0.0            40.0  United-States   >50K  
3     Male        7688.0           0.0            40.0  United-States   >50K  
4   Female           0.0           0.0            30.0  United-States  <=50K  
5     Male           0.0           0.0            30.0  United-States  <=50K  
6     Male           0.0           0.0            40.0  United-States  <=50K  
7     Male        3103.0           0.0            32.0  United-States   >50K  
8   Female           0.0           0.0            40.0  United-States  <=50K  
9     Male           0.0           0.0            10.0  United-States  <=50K  
10    Male        6418.0 

In [16]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9615 entries, 0 to 9614
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   age              9615 non-null   int64  
 1   workclass        9615 non-null   object 
 2   fnlwgt           9615 non-null   int64  
 3   education        9615 non-null   object 
 4   educational-num  9615 non-null   int64  
 5   marital-status   9615 non-null   object 
 6   occupation       9615 non-null   object 
 7   relationship     9615 non-null   object 
 8   race             9614 non-null   object 
 9   gender           9614 non-null   object 
 10  capital-gain     9614 non-null   float64
 11  capital-loss     9614 non-null   float64
 12  hours-per-week   9614 non-null   float64
 13  native-country   9614 non-null   object 
 14  income           9614 non-null   object 
dtypes: float64(3), int64(3), object(9)
memory usage: 1.1+ MB


In [9]:
data[9614:]

age workclass  fnlwgt   education  educational-num marital-status  \
9614   49   Private  123681  Assoc-acdm               12      Separated   

     occupation relationship  race gender  capital-gain  capital-loss  \
9614      Sales    Unmarried  None   None           NaN           NaN   

      hours-per-week native-country income  
9614             NaN           None   None

In [50]:
data = data[:9614]

In [17]:
data.workclass.value_counts()

Private             6574
Self-emp-not-inc     782
Local-gov            609
?                    584
State-gov            429
Self-emp-inc         353
Federal-gov          279
Without-pay            4
Never-worked           1
Name: workclass, dtype: int64

In [18]:
# заполняем неизвестные значения самыми частыми
data.workclass[data.workclass == '?'] = 'Private'

In [19]:
data.occupation.value_counts()

Prof-specialty       1222
Exec-managerial      1213
Craft-repair         1183
Adm-clerical         1079
Sales                1066
Other-service         960
Machine-op-inspct     621
?                     585
Transport-moving      429
Handlers-cleaners     414
Farming-fishing       310
Tech-support          279
Protective-serv       194
Priv-house-serv        57
Armed-Forces            3
Name: occupation, dtype: int64

In [22]:
corr_q = data[data.occupation == '?'].corr() # строим матрицу корреляций объектов, у
                                             # которых место работы '?'

corr_q.style.background_gradient(cmap='coolwarm')

Сильной корреляции ни с каким признаком нет. Так как данных с местом работы "?" много (585), то просто назовем им отдельной категорией "other"

In [23]:
data.occupation[data.occupation == '?'] = 'other'

In [25]:
data['native-country'].value_counts()

United-States                 8681
Mexico                         186
?                              164
Philippines                     51
Germany                         41
Puerto-Rico                     41
India                           34
Canada                          34
El-Salvador                     30
Cuba                            28
England                         21
Poland                          20
Haiti                           20
South                           20
China                           19
Dominican-Republic              18
Portugal                        18
Guatemala                       17
Italy                           15
Jamaica                         15
Ecuador                         13
Columbia                        13
Japan                           12
Peru                            11
Thailand                        10
Nicaragua                        9
Taiwan                           7
Vietnam                          7
Greece              

In [26]:
# заполняем наиболее частым значением
data['native-country'][data['native-country'] == '?'] = 'United-States'

In [54]:
data['hours-per-week'] = data['hours-per-week'].astype('int64')

In [55]:
data.income[data.income == '>50K'] = 0
data.income[data.income == '<=50K'] = 1

In [56]:
data.income.value_counts()

1    7356
0    2258
Name: income, dtype: int64

In [58]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9614 entries, 0 to 9613
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   age              9614 non-null   int64  
 1   workclass        9614 non-null   object 
 2   fnlwgt           9614 non-null   int64  
 3   education        9614 non-null   object 
 4   educational-num  9614 non-null   int64  
 5   marital-status   9614 non-null   object 
 6   occupation       9614 non-null   object 
 7   relationship     9614 non-null   object 
 8   race             9614 non-null   object 
 9   gender           9614 non-null   object 
 10  capital-gain     9614 non-null   float64
 11  capital-loss     9614 non-null   float64
 12  hours-per-week   9614 non-null   int64  
 13  native-country   9614 non-null   object 
 14  income           9614 non-null   object 
dtypes: float64(2), int64(4), object(9)
memory usage: 1.1+ MB


In [57]:
corr = data.corr() # строим матрицу корреляций

corr.style.background_gradient(cmap='coolwarm')

In [59]:
X = data.drop('income', axis=1) # матрица объект-признак
y = data['income'] # целевая переменная

In [76]:
from sklearn.model_selection import train_test_split

Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, test_size=0.25, random_state=42)

In [77]:
from sklearn.preprocessing import MinMaxScaler

numerical = ['age', 'fnlwgt', 'educational-num', 'capital-gain', 'capital-loss', 'hours-per-week']

ss = MinMaxScaler()
ss.fit(Xtrain[numerical])

Xtrain_num = pd.DataFrame(ss.transform(Xtrain[numerical]), columns = numerical)
Xtest_num = pd.DataFrame(ss.transform(Xtest[numerical]), columns = numerical)

Xtrain_num.head()

age    fnlwgt  educational-num  capital-gain  capital-loss  \
0  0.027397  0.056166         0.600000           0.0           0.0   
1  0.054795  0.342985         0.600000           0.0           0.0   
2  0.109589  0.166559         0.533333           0.0           0.0   
3  0.191781  0.164820         0.800000           0.0           0.0   
4  0.219178  0.108657         0.800000           0.0           0.0   

   hours-per-week  
0        0.132653  
1        0.397959  
2        0.500000  
3        0.346939  
4        0.551020

In [79]:
from sklearn.preprocessing import OneHotEncoder

categorical = ['workclass','education','marital-status','occupation', 'relationship', 'race', 'gender', 'native-country']

OH_encoder = OneHotEncoder(handle_unknown='error', sparse=False)
OH_encoder.fit(Xtrain[categorical])

Xtrain_ohe = pd.DataFrame(OH_encoder.transform(Xtrain[categorical]), columns=OH_encoder.get_feature_names_out())
Xtest_ohe = pd.DataFrame(OH_encoder.transform(Xtest[categorical]), columns=OH_encoder.get_feature_names_out())

In [80]:
Xtrain_new = pd.concat([Xtrain_num, Xtrain_ohe], axis=1)
Xtest_new = pd.concat([Xtest_num, Xtest_ohe], axis=1)

In [81]:
Xtrain_new[:10]

age    fnlwgt  educational-num  capital-gain  capital-loss  \
0  0.027397  0.056166         0.600000           0.0           0.0   
1  0.054795  0.342985         0.600000           0.0           0.0   
2  0.109589  0.166559         0.533333           0.0           0.0   
3  0.191781  0.164820         0.800000           0.0           0.0   
4  0.219178  0.108657         0.800000           0.0           0.0   
5  0.287671  0.144820         0.400000           0.0           0.0   
6  0.643836  0.176933         0.600000           0.0           0.0   
7  0.328767  0.170847         0.800000           0.0           0.0   
8  0.534247  0.128320         0.600000           0.0           0.0   
9  0.109589  0.109552         0.800000           0.0           0.0   

   hours-per-week  workclass_Federal-gov  workclass_Local-gov  \
0        0.132653                    0.0                  0.0   
1        0.397959                    0.0                  0.0   
2        0.500000                    0.0                  0.0   
3        0.346939                    0.0                  0.0   
4        0.551020                    0.0                  0.0   
5        0.397959                    0.0                  0.0   
6        0.346939                    0.0                  0.0   
7        0.397959                    0.0                  0.0   
8        0.397959                    0.0                  1.0   
9        0.397959                    0.0                  0.0   

   workclass_Never-worked  workclass_Private  ...  native-country_Portugal  \
0                     0.0                1.0  ...                      0.0   
1                     0.0                1.0  ...                      0.0   
2                     0.0                1.0  ...                      0.0   
3                     0.0                1.0  ...                      0.0   
4                     0.0                1.0  ...                      0.0   
5                     0.0                1.0  ...                      0.0   
6                     0.0                0.0  ...                      0.0   
7                     0.0                1.0  ...                      0.0   
8                     0.0                0.0  ...                      0.0   
9                     0.0                1.0  ...                      0.0   

   native-country_Puerto-Rico  native-country_Scotland  native-country_South  \
0                         0.0                      0.0                   0.0   
1                         0.0                      0.0                   0.0   
2                         0.0                      0.0                   0.0   
3                         0.0                      0.0                   0.0   
4                         0.0                      0.0                   0.0   
5                         0.0                      0.0                   0.0   
6                         0.0                      0.0                   0.0   
7                         0.0                      0.0                   0.0   
8                         0.0                      0.0                   0.0   
9                         0.0                      0.0                   0.0   

   native-country_Taiwan  native-country_Thailand  \
0                    0.0                      0.0   
1                    0.0                      0.0   
2                    0.0                      0.0   
3                    0.0                      0.0   
4                    0.0                      0.0   
5                    0.0                      0.0   
6                    0.0                      0.0   
7                    0.0                      0.0   
8                    0.0                      0.0   
9                    0.0                      0.0   

   native-country_Trinadad&Tobago  native-country_United-States  \
0                             0.0                           1.0   
1                             0.0                           1.0

In [82]:
!pip install catboost -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.6/76.6 MB 9.5 MB/s eta 0:00:00


In [91]:
from catboost import CatBoostClassifier

ytrain = ytrain.astype(int)

gbm = CatBoostClassifier(iterations=300, #сколько деревьев
                          depth=2, #глубина каждого дерева
                          learning_rate=0.1,
                          loss_function='Logloss',
                          eval_metric='AUC',
                          verbose=False)
gbm.fit(Xtrain_new, ytrain)
gbm_preds = gbm.predict(Xtest_new)

In [92]:
ytest = ytest.astype(int)

In [93]:
from sklearn.metrics import roc_auc_score

roc_auc_score(ytest, gbm_preds)

0.7581413904892176

In [94]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression() # объявление модели
# обучение модели

In [ ]:
model.fit(Xtrain_new, ytrain) 

In [96]:
pred_classes = model.predict(Xtest_new) # предсказание классов

In [97]:
from sklearn.metrics import accuracy_score

accuracy_score(ytest, pred_classes)

0.8406821963394343

In [98]:
from sklearn.metrics import roc_auc_score

roc_auc_score(ytest, pred_classes)

0.7390050656682161

In [ ]:
from sklearn.model_selection import GridSearchCV

gbm2 = CatBoostClassifier(iterations = 500,
                          loss_function = 'Logloss',
                          learning_rate = 0.05,
                          eval_metric='AUC',
                          verbose=False)

params = {'max_depth' : [2, 3, 4, 5, 6]}

gs = GridSearchCV(gbm2, params, cv = 5, scoring='roc_auc')


In [114]:
gs.best_estimator_, gs.best_score_, gs.best_params_

(<catboost.core.CatBoostClassifier at 0x7f7213105ac0>,
 0.9214860375117133,
 {'max_depth': 6})

Итоговая оценка roc_auc = 0.9214860375117133